Add Elasticsearch output plugin (#374)

micrictor opened the pull request:
Creates an Elasticsearch output plugin using the `_bulk` API.

In the config file, the GRR Admin sets:
* url - Required; the Elasticsearch endpoint to index the documents at
* token - Optional; The authentication token to include with the indexing requests
* verify_https - Optional; Whether to validate the server certificate. Includes a disclaimer on why this is a bad idea
* index - Optional; The elasticsearch index to place the flow results into; Default "grr-flows"

When executing the flow, the operator can specify:
* tags - Arbitrary tags placed on the flow document(s)
* index - Destination index for the documents, if desired to override the configured index

Note that, as a collateral change, I had to add an optional `indent` parameter to the `json.Dump` method to avoid messing up the line-delimited nature of the Elasticsearch API.

This resolves #374

max-vogler commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,213 @@
+    # Elasticsearch bulk requests are line-deliminated pairs, where the first
+    # line is the index command and the second is the actual document to index
+    split_request = request.split('\n')
+    print(split_request)
```
Please remove the `print()` debugging statement.

max-vogler commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,213 @@
+    print(split_request)
+    update_pairs = [(split_request[i], split_request[i + 1])
+                    for i in range(0, len(split_request), 2)]
+    parsed_pairs = [(json.Parse(i[0]), json.Parse(i[1])) for i in update_pairs]
```
Instead of deconstructing the tuple, parsing the elements, and reconstructing a tuple, could you call `json.Parse` on the whole list in `split_requests`? This might simplify the code a little, e.g.:

```py
split_request = [json.Parse(line) for line in request.split("\n")]
```

max-vogler commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py:
```diff
@@ -0,0 +1,155 @@
+    # Each index operation is two lines, the first defining the index settings,
+    # the second is the actual document to be indexed
+    data = "\n".join(
+      index_command + "\n" + json.Dump(event, indent=None)
```
Please use string formatting (f-strings or `format()`) instead of manual string concatenation, e.g.:

```py
    data = "\n".join(
      f"{index_command}\n{json.Dump(event, indent=None)}"
      for event in events)
```

micrictor replied:
Done - new code is:

```
data = "\n".join([
      "{}\n{}".format(index_command, json.Dump(event, indent=None))
      for event in events])
```

max-vogler reviewed (approved):
Thank you for this amazing pull request! I'm pre-approving, because there are only minor code health changes to be made and I am out of office next week. One of my teammates can merge next week or I can merge the week after when I'm back.

micrictor commented:
Ack on all accounts - debug print is an obvious oopsie. I’ll try to figure
out a... less bad way to decompose the query data. I contemplated making a
custom object just to represent this relationship, but that felt like it
would be taking away from the readability of the test.

I see various python2 interop stuff throughout the code base, so I’ll use
`format` in case there’s someone still relying on that.

On Fri, Apr 9, 2021 at 4:47 AM Max Vogler ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> Thank you for this amazing pull request! I'm pre-approving, because there
> are only minor code health changes to be made and I am out of office next
> week. One of my teammates can merge next week or I can merge the week after
> when I'm back.
> ------------------------------
>
> In
> grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py
> <https://github.com/google/grr/pull/923#discussion_r610555143>:
>
> > +      patcher = mock.patch.object(requests, 'post')
> +
> +    with patcher as patched:
> +      plugin.ProcessResponses(plugin_state, messages)
> +      plugin.Flush(plugin_state)
> +      plugin.UpdateState(plugin_state)
> +
> +    return patched
> +
> +  def _ParseEvents(self, patched):
> +    request = patched.call_args[KWARGS]['data']
> +
> +    # Elasticsearch bulk requests are line-deliminated pairs, where the first
> +    # line is the index command and the second is the actual document to index
> +    split_request = request.split('\n')
> +    print(split_request)
>
> Please remove the print() debugging statement.
> ------------------------------
>
> In
> grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py
> <https://github.com/google/grr/pull/923#discussion_r610555874>:
>
> > +      plugin.ProcessResponses(plugin_state, messages)
> +      plugin.Flush(plugin_state)
> +      plugin.UpdateState(plugin_state)
> +
> +    return patched
> +
> +  def _ParseEvents(self, patched):
> +    request = patched.call_args[KWARGS]['data']
> +
> +    # Elasticsearch bulk requests are line-deliminated pairs, where the first
> +    # line is the index command and the second is the actual document to index
> +    split_request = request.split('\n')
> +    print(split_request)
> +    update_pairs = [(split_request[i], split_request[i + 1])
> +                    for i in range(0, len(split_request), 2)]
> +    parsed_pairs = [(json.Parse(i[0]), json.Parse(i[1])) for i in update_pairs]
>
> Instead of deconstructing the tuple, parsing the elements, and
> reconstructing a tuple, could you call json.Parse on the whole list in
> split_requests? This might simplify the code a little, e.g.:
>
> split_request = [json.Parse(line) for line in request.split("\n")]
>
> ------------------------------
>
> In grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py
> <https://github.com/google/grr/pull/923#discussion_r610559497>:
>
> > +  def _SendEvents(self, events: List[JsonDict]) -> None:
> +    """Uses the Elasticsearch bulk API to index all the events in a single
> +    request.
> +    https://www.elastic.co/guide/en/elasticsearch/reference/7.1/docs-bulk.html
> +    """
> +    if self._token:
> +      headers = {"Authorization": "Basic {}".format(self._token)}
> +    else:
> +      headers = {}
> +
> +    index_command = json.Dump({"index": {"_index": self._index}}, indent=None)
> +
> +    # Each index operation is two lines, the first defining the index settings,
> +    # the second is the actual document to be indexed
> +    data = "\n".join(
> +      index_command + "\n" + json.Dump(event, indent=None)
>
> Please use string formatting (f-strings or format()) instead of manual
> string concatenation, e.g.:
>
>     data = "\n".join(
>       f"{index_command}\n{json.Dump(event, indent=None)}"
>       for event in events)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/grr/pull/923#pullrequestreview-632315856>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTCAGLTDMXI32KB3P2S36DTH3SM5ANCNFSM42UGAN2Q>
> .
>

panhania commented on grr/core/grr_response_core/config/output_plugins.py:
```diff
@@ -63,3 +63,22 @@
+    "See https://requests.readthedocs.io/en/master/user/advanced/.")
+
+config_lib.DEFINE_string("Elasticsearch.index", "grr-flows",
+                         "The index assigned to all submitted events.")
```
Missing trailing blank line (as shown by the GitHub UI).

micrictor replied:
VS Code, why you no lint ): fixed in next rev

panhania commented on grr/core/grr_response_core/lib/util/compat/json.py:
```diff
@@ -67,7 +67,8 @@ def ReadFromPath(filepath: Text) -> Any:
          sort_keys: bool = False,
-         encoder: Optional[Type[Encoder]] = None) -> Text:
+         encoder: Optional[Type[Encoder]] = None,
+         indent: int = 2) -> Text:
```
The new parameter should be documented in the `Args` section.

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py:
```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python
+# Lint as: python3
```
No need for these lines (no one needs nor should run it as a stand-alone executable and nowadays everything is linted in Python 3 mode anyway).

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py:
```diff
@@ -0,0 +1,155 @@
+The specification for the indexing of documents is
+https://www.elastic.co/guide/en/elasticsearch/reference/7.1/docs-index_.html
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import unicode_literals
```
No need for these, we are Python 3-only now.

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py:
```diff
@@ -0,0 +1,155 @@
+
+BULK_OPERATIONS_PATH = "_bulk"
+
+JsonDict = Dict[Text, Any]
```
Could you add a `TODO` comment linking to the [JSON type discussion](https://github.com/python/typing/issues/182)? Once they figure it out, we should fix this one day...

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py:
```diff
@@ -0,0 +1,155 @@
+  args_type = ElasticsearchOutputPluginArgs
+
+  def __init__(self, *args, **kwargs):
+    """See base class."""
```
I think something as simple as _"Initializes the plugin"_ is better than delegating to the base class.

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py:
```diff
@@ -0,0 +1,155 @@
+    # Allow the Flow creator to override the index, fall back to configuration.
+    self._index = self.args.index or config.CONFIG["Elasticsearch.index"]
+
+    if not url:
+      raise ElasticsearchConfigurationError(
+          "Cannot start ElasticsearchOutputPlugin, because Elasticsearch.url"
+          "is not configured. Set it to the base URL of your Elasticsearch"
+          "installation, e.g. 'https://myelasticsearch.example.com:9200'.")
```
I would place right below the `url = config.CONFIG["Elasticsearch.url"]` part—this way you avoid executing code that will never be used because the exception.

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,211 @@
+            patcher=mock.patch.object(requests, 'post', post))
+
+
+def main(argv):
+  test_lib.main(argv)
+
+
+if __name__ == '__main__':
+  app.run(main)
```
You can just do:

```python
if __name__ == '__main__':
  app.run(test_lib.main)
```

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python
+# Lint as: python3
+# -*- encoding: utf-8 -*-
```
These are not needed as well.

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,211 @@
+# Lint as: python3
+# -*- encoding: utf-8 -*-
+"""Tests for Elasticsearch output plugin."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import unicode_literals
```
Ditto.

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,211 @@
+from grr.test_lib import flow_test_lib
+from grr.test_lib import test_lib
+
+KWARGS = 1
```
This warrants a comment.

micrictor replied:
Definitely right - I don't even remember what this is for any more ):

panhania commented on grr/server/grr_response_server/output_plugins/elasticsearch_plugin_test.py:
```diff
@@ -0,0 +1,211 @@
+    plugin, plugin_state = plugin_cls.CreatePluginAndDefaultState(
+        source_urn=source_id, args=plugin_args)
+
+    if patcher is None:
+      patcher = mock.patch.object(requests, 'post')
```
You can look into the `responses` package (used for example [here](https://github.com/google/grr/blob/05d92771c7915de0cccd2f95fdd8a1f2609c7e3c/grr/client/grr_response_client/gcs_test.py#L26-L87)).

micrictor replied:
Probably going to leave this alone for now, unless you think it should be blocking.

panhania reviewed (commented):
Thanks for the pull request! I left some comments in addition to the @max-vogler's review.

micrictor commented:
Pushed in the updates discussed above - mostly removing python2 compatibility stuff, adding docs for the new `indent` parameter, and other minor formatting fixes.

As noted above, I didn't port the tests to use `responses`, and don't really intend on doing so. If that's blocking, this PR will likely stay open for quite some time before I circle back to that (or someone else picks it up)

max-vogler commented:
I suggest merging without the `responses` test refactoring in order to merge this functionality. The test is very close to splunk_plugin_test.py, which uses the same mocking. In the future, one change can refactor both tests. @panhania WDYT?

panhania commented:
Sounds reasonable to me.

max-vogler commented:
Thanks a lot for your contribution @micrictor, well done! If you like, send me a PR that adds your name and email to [ACKNOWLEDGEMENTS](https://github.com/google/grr/blob/master/ACKNOWLEDGEMENTS).
